lib.strings: init escapePythonBytes

Jayman2000 opened the pull request:
This pull request adds a new `lib.strings.escapePythonBytes` function. It’s like `lib.strings.escapeShellArg` but for the Python programming language instead of shell languages. See the commit message for details.

Here are some things that I wasn’t sure about when creating this pull request:

- I wanted to add a test for `lib.strings.escapePythonBytes` to `lib/tests/misc.nix`, but the test that I wrote depends on `pkgs`. At the moment, `lib/tests/misc.nix` doesn’t have any access to `pkgs` so I wasn’t sure what to do. I decided to create a new `lib/tests/python` directory for the test that I wrote because it looked like we were trying to intentionally avoid making `lib/tests/misc.nix` depend on `pkgs`.

- I added `lib/tests/python` to `lib/tests/release.nix` so that that test will run when you run `nix-build lib/tests/release.nix`. I wasn’t sure if I should add `lib/tests/python` to `lib/tests/test-with-nix.nix` so I didn’t. Should I add `lib/tests/python` to `lib/tests/test-with-nix.nix`?

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`
Commit: `7ca06970ba5d5ea5f1583f693bbb7f21bc5cf984`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>extra-container</li>
    <li>nixos-container</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [x] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Jayman2000 committed:
lib.strings: init escapePythonBytes

I often write Nix expressions that produce scripts that have Nix strings
embedded in them. In other words, I often do stuff like this:

  { pkgs }:
  pkgs.writeShellScriptBin "example-script" ''
    printf 'Location of hello: %q\n' ${pkgs.lib.strings.escapeShellArg "${pkgs.hello}"}
  ''

When I find myself wanting to embed a Nix string into a script, I use
lib.strings.escapeShellArg most of the time because I write most of
those script in Bash. Sometimes though, I write those scripts in the
Python programming language so I can’t use lib.strings.escapeShellArg.

The first time that I remember running into this situation was when I
was working on #289653. I had originally tried writing ecwolf’s update
script in Bash, but I found it difficult to write Bash code that would
parse the data that the script would download from debian.drdteam.org. I
then found a Python library that would do most of the parsing for me so
I switched to using the Python programming language, but that meant that
I could no longer use lib.strings.escapeShellArg. I ended up finding a
workaround (passing the Nix string as a command-line argument to the
Python script), but using workarounds is not ideal.

I recently started working on a Disko [1] pull request, and I ran into
the same problem again. I was writing a NixOS test for the functionality
that my pull request adds, and I found myself wanting to embed a Nix
string into a testScript [2]. I looked in Nixpkgs’s nixos/tests
directory to try and find a workaround. I did not find a workaround.
Instead, I found stuff like this (quote from nixos/tests/3proxy.nix):

>     peer0.succeed(
>         "${pkgs.wget}/bin/wget -e use_proxy=yes -e http_proxy=http://192.168.0.2:3128 -S -O /dev/null http://216.58.211.112:9999"
>     )

Code written like that will break if the path to the Nix store contains
certain characters. For example, in the future, someone’s Nix store
might be located at C:\nix\store [3]. In that situation, that previously
quoted chunk of code would turn into something like this:

>     peer0.succeed(
>         "C:\nix\store\<hash>-wget-<version>/bin/wget -e use_proxy=yes -e http_proxy=http://192.168.0.2:3128 -S -O /dev/null http://216.58.211.112:9999"
>     )

The “\n” in that line of code would be interpreted as a newline which is
clearly not what that author of that code intended.

---

This issue seems to come up often enough that we ought to have a proper
solution for it. This change adds a new lib.strings.escapePythonBytes
Nix function so that we don’t have to work around or ignore this problem
anymore.

---

In order to avoid Import From Derivation [4], this change adds some
automatically generated files. I generated those files by building this
Nix expression:

  let
    nixpkgsCommit = "32ac3bd1104493f70a7ad3449abfc4723a1313f3";
    nixpkgsDirectory = builtins.fetchTarball {
      url = "https://github.com/NixOS/nixpkgs/archive/${nixpkgsCommit}.tar.gz";
      sha256 = "08bmaiakl1lygwjgzfxsnb6027vahsvr6vh314kgyx500qbzqqch";
    };
    pkgs = import nixpkgsDirectory { };
  in
  pkgs.runCommandWith
    {
      name = "auto-generated-files-for-escape-python-bytes";
    }
    ''
      mkdir --parents "$out/byte-to-hex-table"
      # We’re skiping 0 here because Nix strings can’t contain null bytes [1].
      #
      # [1]: <https://github.com/NixOS/nix/issues/1307>
      for (( i=1; i < 256; i++))
      do
        printf -v iInHex '%02X' "$i"
        printf "\\x$iInHex" | tee "$out/byte-to-hex-table/$iInHex" >> "$out/all-bytes-except-null.bin"
      done
    ''

---

[1]: <https://github.com/nix-community/disko>
[2]: <https://nixos.org/manual/nixos/unstable/#test-opt-testScript>
[3]: <https://github.com/NixOS/nix/issues/9389>
[4]: <https://github.com/NixOS/nixpkgs/blob/32ac3bd1104493f70a7ad3449abfc4723a1313f3/pkgs/README.md#import-from-derivation>

wolfgangwalther commented on ci/default.nix:
```diff
@@ -117,6 +117,14 @@ let
+            # character encoding for some of them and that most of them don’t
+            # end with a newline.
+            "lib/byte-to-hex-table/??"
+          ];
```
You should be able to create a local `.editorconfig` file in the relevant folder to tell the checker that these things are binary files. No need for an exclusion here, imho.

dotlambda commented on lib/strings.nix:
```diff
@@ -1467,6 +1467,73 @@ rec {
+
+    This function allows you to turn a Nix string value into a Python bytes value rather than a
+    Python string value. This is because the Nix expression language and the Python programming
+    language use the term “string” to refer to different things. In the Nix expression language,
```
What happened here?

Jayman2000 replied:
OK, so when I look at that file locally, I can see that there are Unicode smart quotes around the word string. When I looked at that file on GitHub, it was showing mojibake for me, but then I refreshed the page, and the mojibake went away. Does it look right for you if you refresh the page?

Mic92 committed:
lib.strings.escapePythonBytes: replace filesystem lookup with programmatic generation

The previous implementation used 256 individual files (one per byte value
0x01-0xFF) to build the byte-to-hex lookup table. This filesystem-based
approach was difficult to maintain, required special .gitattributes handling,
and was an unusual pattern in the codebase.

The new implementation generates the lookup table programmatically by
reading from the existing test file (which was previously located at
lib/tests/python/all-bytes-except-null.bin) and using simple arithmetic
to compute hex representations.

This change also moves that test file from lib/tests/python/all-bytes-except-null.bin
to lib/byte-to-hex-table/all-bytes-except-null.bin since it is now used by more
than just the tests.

Co-authored-by: Jason Yundt <jason@jasonyundt.email>

Mic92 commented on lib/strings.nix:
```diff
@@ -1467,6 +1467,73 @@ rec {
+  escapePythonBytes =
+    s:
+    let
+      byteToHexTable = import ./byte-to-hex-table lib;
```
This doesn't require 256 files: https://github.com/NixOS/nixpkgs/compare/master...Mic92:nixpkgs:add-lib.strings.escapePythonBytes?expand=1

MattSturgeon commented on lib/byte-to-hex-table/default.nix:
```diff
@@ -0,0 +1,56 @@
+      high = n / 16;
+      low = n - (high * 16);
+    in
+    builtins.elemAt hexDigits high + builtins.elemAt hexDigits low;
```
Could we use `lib.toHexString` and `lib.fromHexString` here?

Jayman2000 replied:
I just pushed a new version of this pull request that uses `lib.trivial.toHexString`. The new version of this pull request also renames the `toHex` function to `toTwoDigitHexString` in order to highlight the difference between that function and `lib.trivial.toHexString`. That being said, I wasn’t able to find a place to use `lib.trivial.fromHexString`. What did you have in mind for `lib.trivial.fromHexString`?

MattSturgeon commented on lib/byte-to-hex-table/default.nix:
```diff
@@ -0,0 +1,56 @@
+  allBytes = builtins.readFile allBytesFile;
+
+  # Split the string into individual bytes.
+  byteList = builtins.genList (i: builtins.substring i 1 allBytes) (builtins.stringLength allBytes);
```
`lib.stringToCharacters`

MattSturgeon commented on lib/byte-to-hex-table/default.nix:
```diff
@@ -0,0 +1,56 @@
+  byteAttrs = builtins.genList (i: {
+    name = builtins.elemAt byteList i;
+    value = toHex (i + 1);
+  }) 255;
```
I think this can be simplified to `lib.genAttrs`:

```suggestion
  byteAttrs = lib.genAttrs byteList lib.toHexString;
```

MattSturgeon commented on lib/byte-to-hex-table/all-bytes-except-null.bin:
```diff
@@ -0,0 +1,2 @@
+	
+
 !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~��������������������������������������������������������������������������������������������������������������������������������
```
Much better than 255 loose files! 🎉

Jayman2000 committed:
lib.strings.escapePythonBytes: simplify implementation

This commit simplifies the implementation of
lib.strings.escapePythonBytes by incorporating feedback from the
following review comments:

• <https://github.com/NixOS/nixpkgs/pull/458502#discussion_r2516097137>
• <https://github.com/NixOS/nixpkgs/pull/458502#discussion_r2516100489>
• <https://github.com/NixOS/nixpkgs/pull/458502#discussion_r2516108272>
• <https://github.com/NixOS/nixpkgs/pull/458502#discussion_r2516137122>
• <https://github.com/NixOS/nixpkgs/pull/458502#discussion_r2516149214>

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Jayman2000 committed:
lib.strings.escapePythonBytes: avoid many unnecessary escape sequences

See the comment above escapePythonBytes for details.

When making this change, I had considered going farther. I had
considered making it so that lib.strings.escapePythonBytes uses as few
escape sequences as possible. I decided that doing that would be overly
hard and error-prone. If you want to use as few escape sequences as
possible, then you need to choose which of the following types of bytes
literals would result in the least number of escape sequences:

• b'Example'
• b"Example"
• b'''Example'''
• b"""Example"""
• br'Example'
• br"Example"
• br'''Example'''
• br"""Example"""

I decided that writing code that would make that choice correctly was
not worth the effort.

Jayman2000 commented on lib/strings.nix:
```diff
@@ -1467,6 +1467,73 @@ rec {
+    s:
+    let
+      byteToHexTable = import ./byte-to-hex-table;
+      byteToEscapedHex = byte: "\\x${builtins.getAttr byte byteToHexTable}";
```
I opted to incorporate the second suggestion into this pull request. I chose the second suggestion because it looks easier to understand to me.

MattSturgeon commented on lib/tests/python/default.nix:
```diff
@@ -0,0 +1,66 @@
+let
+  allBytesExceptNullPath = ../../byte-to-hex-table/all-bytes-except-null.bin;
+  allBytesExceptNullData = lib.strings.readFile allBytesExceptNullPath;
+  emptyFilePath = ./empty-file.bin;
```
Yes, `import "${nixpkgsDirectory}/lib/tests/python"` is IFD because it is _importing_ `lib/tests/python` from the `nixpkgsDirectory` _derivation_.

The problem with IFD is that eval has to pause while it builds the derivation from which to import. This often leads to major performance issues, which is why it is banned in Nixpkgs.

I guess I didn't spot that this is a file you import from when I suggested `pkgs.emptyFile` (and/or I didn't realize `pkgs.emptyFile` is a derivation).

Jayman2000 committed:
lib.strings.escapePythonBytes: get rid of stringItemTableAttrs

This change simplifies some code by inlining a value. This change was
inspired by this review comment:
<https://github.com/NixOS/nixpkgs/pull/458502#discussion_r2594791810>.

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Jayman2000 commented on lib/byte-to-hex-table/default.nix:
```diff
@@ -0,0 +1,56 @@
+  # byte values programmatically - we can only work with strings that come from
+  # external sources (files, environment variables, etc.).
+  allBytesFile = ./all-bytes-except-null.bin;
+  allBytes = builtins.readFile allBytesFile;
```
That makes sense. I just pushed a new version of this pull request that gets rid of `stringItemTableAttrs`. How do you think the code looks now?

Zocker1999NET reviewed (changes requested):
TL;DR: Thanks ❤️ . But esp. to make the code easier to review later on, I think it can further be improved & simplified simultaneously by extracting the full ASCII table to the top.

Thanks for your work. I found your PR because I was nearly starting implementing a `escapePythonString` function myself. I like your deduction about which characters need to be escaped. (And it was interesting to see that Python does not care about having the lower non-printable ASCII characters in its byte strings 👀.)

But, it took me a really long time understand what you are doing in your code & especially why you need the `all-bytes-except-null.bin` file, because the comment there was not helpful to me. And I assume this is these are the reason why this PR is not merged yet.

Now I got it: You actually need the file to produce a mapping of each byte represented as string into its corresponding integer / ASCII ordinal. And this mapping cannot be represented in Nix natively (as e.g. `charToInt` with [`ascii-table.nix`](https://github.com/NixOS/nixpkgs/blob/1a84b4731d53ca5c47984427017bbc5bfb9cf395/lib/ascii-table.nix) does) because Nix has no `\x01` construct like Python does.

Further the code was hard to read, ironically a little bit because of the long deduction comment, but also because IMO the definition of `byteToStringItemTable` combines two concerns: using `all-bytes-except-null.bin` to map each char/byte to its integer value & then to map each byte to its correctly escaped Python variant.

And separating those concerns does not only make the code more readable IMO, but also reveals the possibility of extending said `charToInt` to support the full range of all bytes (except `^@` of course). (For kind of backwards compatibility, I would suggest to not extend `charToInt` but to introduce a new sister function, just in case some caller relies on `charToInt` failing on non-printable characters.)

That is why I suggest you to further adapt your changes to further simplify and to extract the full ASCII table to the top & optionally adding a `byteToInt` sister function using this full table.

Also, because I did not want to just suggest changes you would then need to apply & debug, I implemented my suggestion in https://github.com/Zocker1999NET/nixpkgs/commit/262e25d08d4c0e5e5b43946a41e9a6603c45d6a1 & verified that `nix-build lib/tests/python` does not only pass but actually evaluate to the exactly same test.

And just as a side note, while it is not technically required, I think I would prefer also escaping all chars with ASCII <= 32 for increasing the readability of the produced Python byte string, for making reading the produced Python code, in case it needs to be debugged, a little bit easier. (See https://github.com/Zocker1999NET/nixpkgs/commit/ed6348757c1f15130288f1de7225bab1a34c4177)

(Feel free to amend my commits into your PR. I hereby allow you to reuse my changes provided in either https://github.com/Zocker1999NET/nixpkgs/commit/262e25d08d4c0e5e5b43946a41e9a6603c45d6a1 & https://github.com/Zocker1999NET/nixpkgs/commit/ed6348757c1f15130288f1de7225bab1a34c4177, or combined, without any kind of attribution.)
